cudamapper/src/index_cache.cuh

```diff
@@ -0,0 +1,224 @@
+    IndexCacheHost(const bool reuse_data,
+                   claragenomics::DefaultDeviceAllocator allocator,
+                   std::shared_ptr<claragenomics::io::FastaParser> query_parser,
+                   std::shared_ptr<claragenomics::io::FastaParser> target_parser,
+                   const std::uint64_t k,
+                   const std::uint64_t w,
+                   const bool hash_representations  = true,
+                   const double filtering_parameter = 1.0,
+                   const cudaStream_t cuda_stream   = 0);
+
+    IndexCacheHost(const IndexCacheHost&) = delete;
+    IndexCacheHost& operator=(const IndexCacheHost&) = delete;
+    IndexCacheHost(IndexCacheHost&&)                 = delete;
+    IndexCacheHost& operator=(IndexCacheHost&&) = delete;
+    ~IndexCacheHost()                           = default;
+
+    /// \brief Discards previously cached query Indices, creates new Indices and copies them to host memory
+    void update_query_cache(const std::vector<IndexDescriptor>& descriptors_of_indices_to_cache);
+
+    /// \brief Discards previously cached target Indices, creates new Indices and copies them to host memory
+    void update_target_cache(const std::vector<IndexDescriptor>& descriptors_of_indices_to_cache);
+
+    /// \brief Copies request Index to device memory
+    /// throws if that index is currently not in cache
+    std::shared_ptr<Index> get_index_from_query_cache(const IndexDescriptor& descriptor_of_index_to_cache);
+
+    /// \brief Copies request Index to device memory
+    /// throws if that index is currently not in cache
+    std::shared_ptr<Index> get_index_from_target_cache(const IndexDescriptor& descriptor_of_index_to_cache);
+
+private:
+    using cache_type_t = std::unordered_map<IndexDescriptor,
+                                            std::shared_ptr<const IndexHostCopyBase>,
+                                            IndexDescriptorHash>;
+
+    enum class CacheToUpdate
+    {
+        QUERY,
+        TARGET
+    };
+
+    /// \brief Discards previously cached Indices, creates new Indices and copies them to host memory
+    /// Uses which_cache to determine if it should be working on query of target indices
+    ///
+    /// If reuse_data_ is true function checks the other cache to see if that index is already in cache
+    void update_cache(const std::vector<IndexDescriptor>& descriptors_of_indices_to_cache,
+                      const CacheToUpdate which_cache);
+
+    cache_type_t query_cache_;
+    cache_type_t target_cache_;
+
+    const bool reuse_data_;
+    claragenomics::DefaultDeviceAllocator allocator_;
+    std::shared_ptr<claragenomics::io::FastaParser> query_parser_;
+    std::shared_ptr<claragenomics::io::FastaParser> target_parser_;
+    const std::uint64_t k_;
+    const std::uint64_t w_;
+    const bool hash_representations_;
+    const double filtering_parameter_;
+    const cudaStream_t cuda_stream_;
```

Comment:
I usually like `const` members because they document the fact that this is a value which is set only once and prevents the developer from accidentally changing them.
In this case I actually explicitly deleted copy and move constructor as I don't expect cached to ever be handled that way and should one ever need they should really think about it and having them manually enable copy/move constructor will make them think once more before doing it.
What do you think?
Also, out of curiosity, why would this prevent the class from being movable? All these values can keep their old values after move

Reply:
Yes, I also used `const` members for the same reason in the past, but now I am ambivalent about them. All of what I don't like about them is related to moving/copying behavior, so since you declared them non-movable non-copyable, I they are fine in this case. In general, here is why I don't like classes with const data members:
a) I cannot default construct them and assign later.
b) They introduce some weird/unusual hybrid between type (like a template instance) and object. This is somewhat a generalization of a) and maybe I just haven't found the right way to think about it yet: if I have objects of a type I expect to be able to copy/move one object the type to another object of the same type, which also means I can create a vector of these objects.
If I introduce real lifetime const data members, I have different "classes" of objects of the same type, but with different constants, where I cannot copy/move objects of one class to objects of any other class. This may be intended, but it is somewhat strange.

One way around both these "shortcomings" would be to define copy and move ctor/assign-op manually and use `const_cast`s there. But that may also be a bit dangerous, because you violate the promised constness for the lifetime of the object and may lead to surprises if you are not fully aware of that when using the object. That's why I just don't use const data members these days.

Reply:
Yes, those are all valid points. I guess it all comes down to what you want to do with the object.

Reply:
yes, indeed. once again. :)